compiler-rt/lib/xray/xray_dso_init.cpp

```diff
@@ -0,0 +1,62 @@
+//===-- xray_init.cpp -------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file is a part of XRay, a dynamic runtime instrumentation system.
+//
+// XRay initialisation logic for DSOs.
+//===----------------------------------------------------------------------===//
+
+#include "sanitizer_common/sanitizer_atomic.h"
+#include "xray_defs.h"
+#include "xray_flags.h"
+#include "xray_interface_internal.h"
+
+using namespace __sanitizer;
+
+extern "C" {
+extern const XRaySledEntry __start_xray_instr_map[] __attribute__((weak))
+__attribute__((visibility("hidden")));
+extern const XRaySledEntry __stop_xray_instr_map[] __attribute__((weak))
+__attribute__((visibility("hidden")));
+extern const XRayFunctionSledIndex __start_xray_fn_idx[] __attribute__((weak))
+__attribute__((visibility("hidden")));
+extern const XRayFunctionSledIndex __stop_xray_fn_idx[] __attribute__((weak))
+__attribute__((visibility("hidden")));
```

Comment:
These symbols are declared `weak` - so can you help me understand how this would behave in different cases like `RTLD_GLOBAL` vs `RTLD_LOCAL`?  How would these be resolved?  Is the behavior acceptable in both cases?  

Say for example I have explicitly linked against liba, libb->libc - all built with `-fxray-enable-shared` and IIRC the loader will load all three libraries with `RTLD_GLOBAL`.  Some time later, my program loads libd (built with `-fxray-enable-shared`) via an explicit `dlopen(..., RTLD_LOCAL)` and the loader resolves one of libd's DT_NEEDED to libe (built with `-fxray-enable-shared`).

In these cases, do we write to these maps, indices such that the patching, unpatching would work as expected?

Can we cover them with tests?

Reply:
How would this symbol resolution behave in a case where the main executable was not built with `-fxray-enable` and only the shared objects were?  If this case is unsupported, can we detect it and log a message or abort()/terminate?

Reply:
The `{start,stop}_xray_instr_map`  symbols are resolved to the begin and end addresses of the `xray_instr_map` ELF section. Same for the `xray_fn_idx` section.

They are using weak linkage so that if, for whatever reason, these sections don't exist, the linker will not fail.
In this case, the XRay runtime will simply not register any patchable sleds. 
I adopted this behavior from the existing `xray_init.cpp`.

Because they are using hidden visibility and each DSO has their own `xray_dso_init.cpp` statically linked, these symbols should resolve correctly regardless of `RTLD_GLOBAL`/`RTLD_GLOBAL`. 

I therefore think that the situation you describe should be handled correctly. 
I will add tests to verify this.

Reply:
To your second point: 

With `-fxray-enable`, are you refering to the general `-fxray-instrument` flag or `-fxray-enable-shared`?

If the main executable was built entirely without `-fxray-instrument`, linking the DSO will fail, as the `__xray_register_dso` and `__xray_deregister_dso` functions could not be resolved.

Otherwise, the XRay runtime will currently always work with instrumented DSOs,  regardless of whether the executable was compiled with `-fxray-enable-shared` or not.
I don't see a reason to change this, since IMO there is no benefit (performance or otherwise) from disabling this feature statically.

I can see practical use cases where one might want to explicitly disable patching DSOs (all or a specific subset).
We could handle this by adding a runtime option, e.g. `xray_patch_dsos=true/false/list of names`.
However, this would probably be best as a separate PR. 

Let me know what you think.

Reply:
I have added a test for the described scenario in [compiler-rt/test/xray/TestCases/Posix/dso-dep-chains.cpp](https://github.com/llvm/llvm-project/pull/90959/files#diff-01ed9b08b391513366a28d78cadb47b46f7b76f8b03163cd1e196e46b9892304).

Reply:
> With -fxray-enable, are you refering to the general -fxray-instrument flag or -fxray-enable-shared?

Yes, the general flag.

> If the main executable was built entirely without -fxray-instrument, linking the DSO will fail, as the __xray_register_dso and __xray_deregister_dso functions could not be resolved.

Ok, right - that makes sense.

> However, this would probably be best as a separate PR.

Agreed - if you decide it's appropriate, good for a follow-up.